### PR TITLE
#2180 - Queued Assessments - Assessment Workflow Enqueuer Scheduler - Start Assessment (Part 1)

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- Created the new queue configuration for the scheduler, executed every 30 seconds, and cleaned up every 24 hours.
- Removed from the code the points where the queue to start the workflow was invoked. From now on only the scheduler will invoke the workflow start queue.
    - Method `startAssessment` removed since it is no longer needed.
    - Method `hasIncompleteAssessment` and `assertAllAssessmentsCompleted` removed since there is no purpose anymore. 
    - Multiple workflows can now be saved and the scheduler will coordinate their execution.
- `StartApplicationAssessment` queue can now be started without providing the `workflowName`.
- Created a migration to update the assessment status based on `assessment_workflow_id`, and `assessment_data`. The rollback is less meaningful and it was created to try to bring the column back to its original state where the column was set with its default value `Submitted`.

TODO:
- E2E tests
- Logs
    - Move `ProcessSummaryResult` from integrations to `lib` to allow its use outside integrations.

### CODE REVIEW

File: sources/packages/backend/apps/db-migrations/src/main.ts [hunk 1]
```diff
@@ -13,7 +13,7 @@ import { ormConfig } from "./data-source";
     await dataSource.query(`SET search_path TO ${ormConfig.schema}, public;`);
     await dataSource.query(`SET SCHEMA '${ormConfig.schema}';`);
     console.info(`**** Running migration ****`);
-    await dataSource.runMigrations();
+    await dataSource.runMigrations({ transaction: "each" });
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/db-migrations/src/main.ts:16]: What does `{ transaction: "each" }` mean?

File: sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts [hunk 2]
```diff
@@ -1,21 +1,24 @@
...
+import { InjectRepository } from "@nestjs/typeorm";
 
 @Injectable()
-export class ApplicationService extends RecordDataModelService<Application> {
+export class ApplicationService {
   constructor(
-    dataSource: DataSource,
     private readonly configService: ConfigService,
-  ) {
-    super(dataSource.getRepository(Application));
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts:17]: Why did you remove the "default" repo here?

**Review by andrepestana-aot — COMMENTED** [review]
Good job. Left some questions.

File: sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Rollback-update-student-assessment-status.sql [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+ Rollback assessments status updates.
+ The workflow is already changing the status to 'In progress' and 'Completed',
+ anything different can be considered as the default column value 'Submitted'.
+ */
+UPDATE
+  sims.student_assessments
+SET
+  student_assessment_status = 'Submitted'
+WHERE
+  student_assessment_status NOT IN ('In progress', 'Completed')
```
Comment (ann-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Rollback-update-student-assessment-status.sql:11]: I just wanted to double-check, During rollback, should we also consider `Cancelled` status?

File: sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    private readonly applicationService: ApplicationService,
+    @InjectQueue(QueueNames.StartApplicationAssessment)
+    private readonly startAssessmentQueue: Queue<StartAssessmentQueueInDTO>,
+  ) {}
+
+  /**
+   * Search applications with pending assessments to be processed by the assessment workflow.
+   * If no other assessment is being processed for that application the oldest pending
+   * assessment will be queue for start.
+   * @returns
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts:32]: minor, return comment

File: sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+          throw new Error(
+            "Student assessment update did not affected any records.",
+          );
+        }
+      });
+      result.summary.push(
+        `Adding assessment to queue ${QueueNames.StartApplicationAssessment}.`,
+      );
+      await this.startAssessmentQueue.add({ assessmentId: nextAssessment.id });
+      result.summary.push(`Assessment queued for start.`);
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts:109]: double quotes?

File: sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts [hunk 6]
```diff
@@ -54,4 +57,56 @@ export class ApplicationService extends RecordDataModelService<Application> {
...
+  /**
+   * Finds all applications with some pending student assessment to be
+   * processed by the workflow, ignoring applications that already have
+   * some workflows in progress or were already queued for execution.
+   * All pending student assessments will be returned ordered by its creation
+   * date to allow the select of the next one to be executed (usually only one
+   * record would be expected).
+   * @returns applications with pending assessments to be executed.
+   */
+  async getApplicationsToStartAssessments(): Promise<Application[]> {
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts:70]: just to refer to my memory, if a student has multiple applications waiting for an assessment, was our plan to process the application in parallel?

**Review by ann-aot — COMMENTED** [review]
Good work @andrewsignori-aot 👍 Added some minor comments and questions. I went through the PR description, I would like to have a walk-through/quick sync-up for the migration issue.

File: sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts:17]: It was a minor refactor. Based on past conversations the idea was to prefer the syntax using `@InjectRepository(Application)` instead of the one being refactored (we can review the benefits if needed).
The services under queue-consumers are following it and the idea is follow on every new class.

File: sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Rollback-update-student-assessment-status.sql (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Rollback-update-student-assessment-status.sql:11]: There is nothing right now in the solution that would change the assessment status to `Cancelled`. It will be introduced in upcoming tickets.

File: sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts:70]: Different applications will be processed in parallel, and assessments from the same application will be processed in order.

File: sources/packages/backend/apps/db-migrations/src/main.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/db-migrations/src/main.ts:16]: Assuming that there are multiple migrations to be executed to have the DB updated, `each` will execute all the commands of the same migration file in a single transaction committing it every time.
Out-of-box behavior would be, to open a transaction, execute all migrations, and commit them all.

File: sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts]: Comment added.

File: sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts]: Changed.

File: sources/packages/backend/apps/api/src/services/education-program-offering/education-program-offering.service.ts [hunk 7]
```diff
@@ -1202,14 +1202,6 @@ export class EducationProgramOfferingService extends RecordDataModelService<Educ
         });
         promises.push(deleteAssessmentPromise);
       }
-      if (application.applicationStatus === ApplicationStatus.Completed) {
-        const startAssessmentPromise =
-          this.workflowClientService.startApplicationAssessment(
-            application.workflowName,
-            application.currentAssessment.id,
-          );
-        promises.push(startAssessmentPromise);
```
Comment (guru-aot) [inline sources/packages/backend/apps/api/src/services/education-program-offering/education-program-offering.service.ts:1211]: As these promises are not part of the async operation, do we still need the restrict the amount of parallel operations with respect to the cpu limit? Or does it need a refactor?

File: sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts [hunk 8]
```diff
@@ -352,10 +352,6 @@ export class StudentAppealService extends RecordDataModelService<StudentAppeal>
       );
     }
 
-    await this.studentAssessmentService.assertAllAssessmentsCompleted(
```
Comment (guru-aot) [inline sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts:355]: Please remove StudentAssessmentService from the constructor

File: sources/packages/backend/apps/api/src/services/student-scholastic-standings/student-scholastic-standings.service.ts [hunk 9]
```diff
@@ -107,10 +107,6 @@ export class StudentScholasticStandingsService extends RecordDataModelService<St
         INVALID_OPERATION_IN_THE_CURRENT_STATUS,
       );
     }
-    // Check if any assessment already in progress for this application.
-    await this.studentAssessmentService.assertAllAssessmentsCompleted(
```
Comment (guru-aot) [inline sources/packages/backend/apps/api/src/services/student-scholastic-standings/student-scholastic-standings.service.ts:111]: https://github.com/bcgov/SIMS/pull/2285#discussion_r1323285401

File: sources/packages/backend/apps/api/src/services/education-program-offering/education-program-offering.service.ts (hunk 7 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/services/education-program-offering/education-program-offering.service.ts:1211]: The processes are already restricted by the amount of CPU cores. Line 1211 takes care only of some leftover (always lesser than the total CPU) cores.
Besides that, the only intention of this method right now is to trigger cancel operations, which will no longer be needed once the upcoming tickets for the queue consumers are done.

File: sources/packages/backend/apps/db-migrations/src/sql/Queue/Add-assessment-workflow-enqueuer-queue.sql [hunk 10]
```diff
@@ -0,0 +1,11 @@
+INSERT INTO
+  sims.queue_configurations(queue_name, queue_configuration)
+VALUES
+  (
+    'assessment-workflow-enqueuer',
+    '{
+        "dashboardReadonly": false,
```
Comment (guru-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/Queue/Add-assessment-workflow-enqueuer-queue.sql:7]: Why we are not having a retry here, just for my understanding.

File: sources/packages/backend/apps/api/src/services/student-scholastic-standings/student-scholastic-standings.service.ts (hunk 9 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/services/student-scholastic-standings/student-scholastic-standings.service.ts:111]: Done!

File: sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts (hunk 8 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts:355]: Done!

File: sources/packages/backend/apps/db-migrations/src/sql/Queue/Add-assessment-workflow-enqueuer-queue.sql (hunk 10 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/db-migrations/src/sql/Queue/Add-assessment-workflow-enqueuer-queue.sql:7]: Because it runs every 30 seconds 😉 
image

File: sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts [hunk 11]
```diff
@@ -54,4 +57,56 @@ export class ApplicationService extends RecordDataModelService<Application> {
...
+    // Sub query to validate if an application has assessment already being
+    // processed by the workflow.
+    const inProgressStatusesExistsQuery = this.studentAssessmentRepo
+      .createQueryBuilder("studentAssessment")
+      .select("1")
```
Comment (guru-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts:75]: I believe the understanding here for the select(1) is, the applicationId will not be returned if either of the 2 inProgressStatuses are present for the particular applicationId right?

File: sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Rollback-update-student-assessment-status.sql (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Rollback-update-student-assessment-status.sql:11]: Sorry, my bad, I completely misunderstood the point, and thanks for the conversation. I will add the `Cancelled` status to the rollback script.

**Review by ann-aot — APPROVED** [review]
Good work @andrewsignori-aot 👍 Thanks for the explanation.

File: sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts (hunk 11 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts:75]: Yes, if the application has some assessment already in progress we will wait for it to be finished to start the next one.

File: sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Update-student-assessment-status.sql [hunk 12]
```diff
@@ -31,4 +31,10 @@ FROM
   sims.applications
 WHERE
   application_id = sims.applications.id
-  AND sims.applications.application_status IN ('Cancelled', 'Overwritten')
\ No newline at end of file
+  AND sims.applications.application_status NOT IN (
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Update-student-assessment-status.sql:34]: 👍

File: sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Update-student-assessment-status.sql [hunk 13]
```diff
@@ -0,0 +1,40 @@
+-- Updates all assessments associated with active applications.
+UPDATE
+  sims.student_assessments sa
+SET
+  student_assessment_status = (
+    CASE
+      WHEN sa.assessment_workflow_id IS NULL THEN 'Submitted'
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Update-student-assessment-status.sql:7]: Submitted is default status of the column, should we need a condition for that?

File: sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Rollback-update-student-assessment-status.sql [hunk 14]
```diff
@@ -0,0 +1,11 @@
+/*
+ Rollback assessments status updates.
+ The workflow is already changing the status to 'In progress' and 'Completed',
+ anything different can be considered as the default column value 'Submitted'.
+ */
+UPDATE
+  sims.student_assessments
+SET
+  student_assessment_status = 'Submitted'
+WHERE
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Rollback-update-student-assessment-status.sql:10]: The rollback may be just setting back the default value to column ('Submitted') without condition.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/start-application-assessment.processor.ts [hunk 15]
```diff
@@ -3,18 +3,30 @@ import { Job } from "bull";
...
 export class StartApplicationAssessmentProcessor {
-  constructor(private readonly workflowClientService: WorkflowClientService) {}
+  constructor(
+    private readonly workflowClientService: WorkflowClientService,
+    private readonly studentAssessmentService: StudentAssessmentService,
+  ) {}
 
   @Process()
   async startAssessment(job: Job<StartAssessmentQueueInDTO>) {
+    let workflowName = job.data.workflowName;
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/start-application-assessment.processor.ts:20]: ❤️

File: sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Rollback-update-student-assessment-status.sql (hunk 14 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Rollback-update-student-assessment-status.sql:10]: It can, but it would be changing possible 'In progress' and 'Completed' statuses already set by the previous ticket already merged.
Between keeping and removing the where both would not be able to put the records back to their original state.
Based on the fact that neither would be perfect I just picked one.

File: sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts [hunk 16]
```diff
@@ -54,4 +57,56 @@ export class ApplicationService extends RecordDataModelService<Application> {
...
+      .where("studentAssessment.application.id = application.id")
+      .andWhere(
+        "studentAssessment.studentAssessmentStatus IN (:...inProgressStatuses)",
+      )
+      .getQuery();
+    return (
+      this.applicationRepo
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts:82]: Just wondering, this query could have come from student assessments instead of applications right? so that it will just return the assesments(assessment id) which needs to be processed.

File: sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Update-student-assessment-status.sql (hunk 13 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Update-student-assessment-status.sql]: I remove it. There is just a small drawback related to garbage data that can exit on local DBs but I did not consider that a good reason to keep it.

File: sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+        result.summary.push(
+          `Associating application currentProcessingAssessment as assessment id ${nextAssessment.id}.`,
+        );
+        const applicationUpdateResult = await entityManager
+          .getRepository(Application)
+          .update(application.id, {
+            currentProcessingAssessment: { id: nextAssessment.id },
+          });
+        if (!applicationUpdateResult.affected) {
+          throw new Error("Application update did not affected any records.");
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts:89]: The error message could be more specific e.g. currentProcessingAssessment id was already set or incorrect application id. something like that.

File: sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+          `Updating assessment status to ${StudentAssessmentStatus.Queued}.`,
+        );
+        const assessmentUpdateResults = await entityManager
+          .getRepository(StudentAssessment)
+          .update(nextAssessment.id, {
+            studentAssessmentStatus: StudentAssessmentStatus.Queued,
+          });
+        if (!assessmentUpdateResults.affected) {
+          throw new Error(
+            "Student assessment update did not affected any records.",
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts:101]: https://github.com/bcgov/SIMS/pull/2285/files#r1323461343

File: sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts (hunk 16 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts:82]: True. It was originally designed to allow the later update of the application + assessment together but it did not work as expected and I ended up doing the individual updates in a transaction.
Right now the only benefit is that it makes it pretty easy to have the assessments grouped to allow the processing and also pick the next assessment to be processed if multiple exists under the same application.

**Comment by dheepak-aot** [comment]
Great job with creating the queue to process assessments. Great to see the depth of analysis and considerations involved in the effort. Thanks for explaining the issue with migrations. Please have a look at the comments.

File: sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts (hunk 17 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts:89]: That log itself may not provide the entire picture, but please check below the logs around it. As shown below there is some additional info that would make it easier to have it traced.

About the suggestions:
- "currentProcessingAssessment id was already set" would not be accurate.
- "incorrect application id" can be checked using the additional logs also.

Queueing next pending assessment for application id 123.
Found 2 pending assessment(s). Queueing assessment 999.
Associating application currentProcessingAssessment as assessment id 999.
(...)
Application update did not affected any records.

File: sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts (hunk 18 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts:101]: https://github.com/bcgov/SIMS/pull/2285#discussion_r1323470028

**Review by andrepestana-aot — APPROVED** [review]
Looks good! Thanks for the explanation.

File: sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts (hunk 17 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts:89]: By the way, planning to do some logging improvement in the next PR.

**Review by dheepak-aot — APPROVED** [review]
Thanks for making the changes 👍

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrewsignori-aot
